### PR TITLE
Add annotations for nullability and length

### LINKED ISSUE #114: Is it possible to make the lengths of arrays a bit more explicit
I'm writing some Haskell bindings to VulkanMemoryAllocator, and it would be really nice to have information on the lengths of arrays being passed into commands in VMA, providing the same information that the `len` attribute in the Vulkan xml spec does.

For example in the below definition, perhaps we could change: `@param[out] pAllocations Pointer to array that will be filled with handles to created allocations.` to `@param[out] pAllocations Pointer to array of allocationCount VmaAllocation values that will be filled with handles to created allocations.`

```C
/** \brief General purpose memory allocation for multiple allocation objects at once.

@param allocator Allocator object.
@param pVkMemoryRequirements Memory requirements for each allocation.
@param pCreateInfo Creation parameters for each alloction.
@param allocationCount Number of allocations to make.
@param[out] pAllocations Pointer to array that will be filled with handles to created allocations.
@param[out] pAllocationInfo Optional. Pointer to array that will be filled with parameters of created allocations.

You should free the memory using vmaFreeMemory() or vmaFreeMemoryPages().

Word "pages" is just a suggestion to use this function to allocate pieces of memory needed for sparse binding.
It is just a general purpose allocation function able to make multiple allocations at once.
It may be internally optimized to be more efficient than calling vmaAllocateMemory() `allocationCount` times.

All allocations are made using same parameters. All of them are created out of the same memory pool and type.
If any allocation fails, all allocations already made within this function call are also freed, so that when
returned result is not `VK_SUCCESS`, `pAllocation` array is always entirely filled with `VK_NULL_HANDLE`.
*/
VMA_CALL_PRE VkResult VMA_CALL_POST vmaAllocateMemoryPages(
    VmaAllocator allocator,
    const VkMemoryRequirements* pVkMemoryRequirements,
    const VmaAllocationCreateInfo* pCreateInfo,
    size_t allocationCount,
    VmaAllocation* pAllocations,
    VmaAllocationInfo* pAllocationInfo);
```

Putting this in the documentation would not be the end of the world, but better yet would be in an attribute on the member/parameter like `__attribute__ ((length (allocationCount)))`, this may be a little too *out there* though. This would be behind a macro like `LEN_ANN(allocationCount)` to avoid tripping `-Wattributes`.

Optionality information I *think* is complete in the documentation already, but I'm not certain.

### PR DESCRIPTION
**Pull request by expipiplus1**
These are only present on the declarations in the C API.

New macros
- `VMA_NULLABLE`
- `VMA_NOT_NULL`
- `VMA_NULLABLE_NON_DISPATCHABLE`
- `VMA_NOT_NULL_NON_DISPATCHABLE`
- `VMA_LEN_IF_NOT_NULL`

On clang `VMA_NOT_NULL` and `VMA_NULLABLE` are defined as `_Nonnull` and
`_Nullable`. Clang will use these annotations to generate diagnostics
when these functions or structs are incorrectly used.

`VMA_NULLABLE_NON_DISPATCHABLE` and `VMA_NOT_NULL_NON_DISPATCHABLE` are
not defined on platforms where Vulkan's non-dispatchable handles are not
pointers; on platforms where they are, these macros are defined to be
`VMA_NULLABLE` and `VMA_NOT_NULL` respectively.

`LEN_IF_NOT_NULL` is not defined for any compiler. It is included to
mirror the `len` attribute in the Vulkan XML spec in order to aid in
development of other language bindings for this library.

It was tempting to use C's array syntax (`foo(int len, int arr[len])`)to
specify lengths, however this is not possible to use with optional
pointers, nor is it possible to use inside structs. For the sake of
consistency only `VMA_LEN_IF_NOT_NULL` is used (additionally neither
clang not gcc generate warnings for function misuse when the array
length specifier is anything but an integer literal).

- Also correct a couple of typos
- Also add assert on ppName in vmaGetPoolName

Closes #114

### CODE REVIEW

File: src/vk_mem_alloc.h [hunk 1]
```diff
@@ -3415,12 +3472,12 @@ typedef struct VmaDefragmentationInfo2 {
...
     */
-    VkCommandBuffer commandBuffer;
+    VkCommandBuffer VMA_NULLABLE commandBuffer;
 } VmaDefragmentationInfo2;
 
 typedef struct VmaDefragmentationPassMoveInfo {
-    VmaAllocation allocation;
-    VkDeviceMemory memory;
+    VmaAllocation VMA_NOT_NULL allocation;
+    VkDeviceMemory VMA_NOT_NULL_NON_DISPATCHABLE memory;
```
Comment (expipiplus1) [inline src/vk_mem_alloc.h:3480]: Are the annotations here correct?

File: src/vk_mem_alloc.h [hunk 2]
```diff
@@ -3430,7 +3487,7 @@ To be used with function vmaBeginDefragmentationPass().
 */
 typedef struct VmaDefragmentationPassInfo {
     uint32_t moveCount;
-    VmaDefragmentationPassMoveInfo* pMoves;
+    VmaDefragmentationPassMoveInfo* VMA_NOT_NULL pMoves VMA_LEN_IF_NOT_NULL(moveCount);
```
Comment (expipiplus1) [inline src/vk_mem_alloc.h:3490]: Is the nullability annotation here correct?

**Comment by expipiplus1** [comment]
Thanks! This is excellent! There are some preliminary haskell bindings generated using these annotations here: https://github.com/expipiplus1/vulkan/blob/vma/VulkanMemoryAllocator/src/Graphics/VulkanMemoryAllocator.hs 

I'll upload these to Hackage after giving them a test drive.

**Comment by expipiplus1** [comment]
@adam-sawicki-a I just noticed that I had included the rejected `Add const qualifier to handles where possible` commit in this PR. You might want to revert that if that's still considered rejected! Sorry!

**Comment by adam-sawicki-a** [comment]
Thanks for this quality contribution!

Including #117 is fine - I decided I will eventually merge it.

Should `VMA_LEN_IF_NOT_NULL` have its parameter as string "" or not? You mixed this in your code.

**Comment by expipiplus1** [comment]
Yeah, it can be any expression and I couldn't think of a better way of saying `VkPhysicalDeviceMemoryProperties::memoryHeapCount`. Perhaps the comment here could be improved https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/pull/115/commits/08bc59bee51b3cc8eb11a43595ec17e66f1346b2#diff-44540aefde48c442bbf9b6a5c0fc3fdeR2002 to mention say that in this case it is passed as a string to be parsed by whatever program is consuming this information.

Another way of putting it would be `VMA_LEN_IF_NOT_NULL(VkPhysicalDeviceMemoryProperties.memoryHeapCount)` but I think that might obscure the meaning.

**Comment by adam-sawicki-a** [comment]
Let me know when your Haskell binding is finished or in Beta stage and I will add it to the list in VMA README.md.

**Comment by adam-sawicki-a** [comment]
I made some fixes in your code. Please take a look and see if they are correct (e.g. `VMA_NOT_NULL` -> `VMA_NOT_NULL_NON_DISPATCHABLE` for `VkInstance`).

**Comment by expipiplus1** [comment]
Thanks for double checking! I made a few comments.

**Comment by adam-sawicki-a** [comment]
OK, fixed. Thanks again!

**Pull request merged**